Sprint 05

VladislavMikhaylov opened the pull request.

VladislavMikhaylov committed:
Started sprint_05

GithubDesktop push test

VladislavMikhaylov committed:
sprint_05_in_progress

-создал фабрику вопросов
-создал алерту с результатами

VladislavMikhaylov committed:
sprint_05_push_for_review

-добавил подсчет статистики
-добавил сохранение результатов лучшей игры
-добавил сохранение значения всех сыгранных раундов

VladislavMikhaylov committed:
sprint_05_push_for_review

-добавил подсчет статистики
-добавил сохранение результатов лучшей игры
-добавил сохранение значения всех сыгранных раундов

VladislavMikhaylov committed:
sprint_05_push_for_review_fixed

-исправил среднюю точность

So1ovyov commented on MovieQuiz/Models/AlertModel.swift:
```diff
@@ -0,0 +1,8 @@
+    let title: String
+    let message: String
+    let buttonText: String
+    var completion: () -> Void
```
Тут нет необходимости в переменной, давай использовать константу:
```
let completion: () -> Void
```

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,41 +1,33 @@
+    private var questionFactory: QuestionFactoryProtocol = QuestionFactory()
+    private var currentQuestion: QuizQuestion?
+    private var alertPresenter : AlertPresenterProtocol?
+    private var statisticService: StatisticService  = StatisticService()
```
Предполагается, что ты будешь использовать переменную под протоколом `StatisticServiceProtocol`, а у тебя тип `StatisticService` с имплементацией `StatisticService`, что не соответствует ТЗ.

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -63,29 +59,52 @@ final class MovieQuizViewController: UIViewController {
         yesButton.isEnabled = isEnabled
```
Предлагаю использовать универсальный метод для отключения/включения сразу двух кнопок:
```
    private func changeStateButton(isEnabled: Bool) {
        noButton.isEnabled = isEnabled
        yesButton.isEnabled = isEnabled
    }
```

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -63,29 +59,52 @@ final class MovieQuizViewController: UIViewController {
+            let recordTime = statisticService.bestGame.date.dateTimeString
+            let accuracy = String(format: "%.2f", statisticService.totalAccuracy)
+            
+            let text = "Ваш результат: \(correctAnswers)/\(questionsAmount)\nКоличество сыгранных квизов: \(totalCount)\nРекорд: \(currentRecord) (\(recordTime))\nСредняя точность: \(accuracy)%"
```
Можно использовать [многострочный вариант строки](https://www.swiftbysundell.com/tips/multiline-string-literal-tips-and-tricks/), чтобы строчка кода умещалась в столбец 120-180 символов:

```
let textDog = """
╥━━━━━━━━╭━━╮━━┳
╢╭╮╭━━━━━┫┃▋▋━▅┣
╢┃╰┫┈┈┈┈┈┃┃┈┈╰┫┣
╢╰━┫┈┈┈┈┈╰╯╰┳━╯┣
╢┊┊┃┏┳┳━━┓┏┳┫┊┊┣
╨━━┗┛┗┛━━┗┛┗┛━━┻
"""
```

So1ovyov commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,126 @@
+import UIKit
+
+class QuestionFactory: QuestionFactoryProtocol {
```
Давай добавим модификатор `final` для этого класса. Он необходим для оптимизации и скорости проекта, также не позволит наследоваться от данного класса.

So1ovyov commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,126 @@
+}
+
+
+/*
```
Мок данные можно удалить так как они хранятся в массиве `[QuizQuestion]`

VladislavMikhaylov committed:
sprint_05_fix

- удалил mock данные
- научился разбивать код на строчки
- добавил универсальный метод откл./вкл. сразу двух кнопок
- исправил критическое замечание, в соответствии с ТЗ

VladislavMikhaylov commented:
Максим, добрый день! Спасибо за ревью, за многострочный вариант написания кода отдельное спасибо! Отправил на ревью обновленную версию
